agent/ebs/watcher_test.go

```diff
@@ -300,6 +304,104 @@ func TestHandleDuplicateEBSAttachment(t *testing.T) {
+					taskresourcevolume.TestFileSystem, gomock.Any(), gomock.Any(),
+					gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).Times(1)
+			},
+			expectedErrorStubs:       nil,
+			expectedAttachmentStatus: attachment.AttachmentAttached,
+		},
+		{
+			name: "StageAll returns expected error when CSI NodeStage call fails",
+			setTaskStateExpectations: func(ts *mock_dockerstate.MockTaskEngineState) {
+				ebsResourceAttachment := getEBSResourceAttachment()
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(ebsResourceAttachment, true).Times(2)
+			},
+			setCSIClientExpectations: func(mcc *mock_csiclient.MockCSIClient) {
+				mcc.EXPECT().NodeStageVolume(gomock.Any(),
+					taskresourcevolume.TestVolumeId, gomock.Any(),
+					filepath.Join(hostMountDir, taskresourcevolume.TestSourceVolumeHostPath),
+					taskresourcevolume.TestFileSystem, gomock.Any(), gomock.Any(),
+					gomock.Any(), gomock.Any(), gomock.Any()).Return(errors.New("test")).Times(1)
+			},
+			expectedErrorStubs:       []string{"Failed to initialize EBS volume ID"},
+			expectedAttachmentStatus: attachment.AttachmentNone,
+		},
+		{
+			name: "StageAll returns expected error when already attached",
+			setTaskStateExpectations: func(ts *mock_dockerstate.MockTaskEngineState) {
+				ebsResourceAttachment := getEBSResourceAttachment()
+				ebsResourceAttachment.SetAttachedStatus()
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(ebsResourceAttachment, true).Times(2)
+			},
+			// we won't call NodeStage here
+			setCSIClientExpectations: func(mcc *mock_csiclient.MockCSIClient) {},
+			expectedErrorStubs:       nil,
+			expectedAttachmentStatus: attachment.AttachmentAttached,
+		},
+		{
+			name: "StageAll fails with missing volume ID",
+			setTaskStateExpectations: func(ts *mock_dockerstate.MockTaskEngineState) {
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(nil, false).Times(1)
+				ebsResourceAttachment := getEBSResourceAttachment()
+				ts.EXPECT().GetEBSByVolumeId(taskresourcevolume.TestVolumeId).Return(ebsResourceAttachment, true).Times(1)
+			},
+			setCSIClientExpectations: func(mcc *mock_csiclient.MockCSIClient) {},
+			expectedErrorStubs:       []string{"Unable to find EBS volume with volume ID"},
+			expectedAttachmentStatus: attachment.AttachmentNone,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			defer mockCtrl.Finish()
+
+			ctx := context.Background()
+
+			taskEngineState := mock_dockerstate.NewMockTaskEngineState(mockCtrl)
+			tc.setTaskStateExpectations(taskEngineState)
+			mockDiscoveryClient := mock_ebs_discovery.NewMockEBSDiscovery(mockCtrl)
+			mockTaskEngine := mock_engine.NewMockTaskEngine(mockCtrl)
+			mockTaskEngine.EXPECT().GetDaemonTask(md.EbsCsiDriver).Return(nil).AnyTimes()
+			mockTaskEngine.EXPECT().GetDaemonManagers().Return(nil).AnyTimes()
```

Comment:
I think `StageAll` does not interact with task engine at all? I am curious why we need these expectations.

Reply:
will clean these up.